### PR TITLE
Local config

### LINKED ISSUE #49: Local config file support
Command launcher today load config file by looking at:
1. the file from env variable $<APP_NAME>_CONFIG_PATH
2. If not found, use $HOME/.<APP_NAME>/config.json

We should allow a third way to load a local config file by searching for <APP_NAME>.json file from the current folder to all its parents. If not found, use the default one. 

It will allow us to have several advantages, for example, having different "profiles" of packages for different usages. Ping package versions for CI purpose, etc.  

The new behaviors will be:
1. the file from env variable $<APP_NAME>_CONFIG_PATH
2. search for <APP_NAME>.json from current folder until the root 
3. If not found, use $HOME/.<APP_NAME>/config.json

### PR DESCRIPTION
**Pull request by bhou-crto**
close #49

### CODE REVIEW

File: internal/config/load.go [hunk 1]
```diff
@@ -87,6 +111,34 @@ func initDefaultConfigFile() {
...
+		return "", false
+	}
+}
+
+func hasConfigFile(configRootPath string, configFileName string) bool {
+	_, err := os.Stat(filepath.Join(configRootPath, configFileName))
+	if err == nil {
+		return true
+	}
+	if os.IsNotExist(err) {
```
Comment (gmonceyron) [inline internal/config/load.go:135]: If the default behavior is to return false, I am not sure this check is necessary

File: internal/config/load.go [hunk 2]
```diff
@@ -17,15 +18,38 @@ import (
...
+	wd, _ := os.Getwd()
 	cfgFile := os.Getenv(appCtx.ConfigurationFileEnvVar())
+	localCftFileName := fmt.Sprintf("%s.json", appCtx.AppName())
 	if cfgFile != "" {
+		configMetadata.Reason = fmt.Sprintf("from environment variable: %s", appCtx.ConfigurationFileEnvVar())
+		configMetadata.File = cfgFile
+
 		viper.SetConfigFile(cfgFile)
+	} else if localCfgPath, found := findLocalConfig(wd, localCftFileName); found {
+		configMetadata.Reason = "found config file from working dir or its ancestors"
```
Comment (gmonceyron) [inline internal/config/load.go:44]:
```suggestion
		configMetadata.Reason = "found config file from working dir or its parents"
```

File: internal/config/load.go [hunk 3]
```diff
@@ -16,15 +17,38 @@ import (
...
+		configMetadata.File = cfgFile
+
 		viper.SetConfigFile(cfgFile)
+	} else if localCfgPath, found := findLocalConfig(wd, localCftFileName); found {
+		configMetadata.Reason = "found config file from working dir or its parents"
+		configMetadata.File = localCfgPath
+
+		viper.SetConfigFile(localCfgPath)
 	} else {
+		configMetadata.Reason = fmt.Sprintf("use default config file from app home %s", AppDir())
```
Comment (lhutyra) [inline internal/config/load.go:49]: It's 3 times AppDir() invocation. AppDir consists of several operation(reading env data, tracing), which seem to be redundant just to put directory in these 3 location below

File: internal/config/load.go (hunk 1 above)
Comment (bhou-crto) [reply internal/config/load.go]: Fixed

File: internal/config/load.go [hunk 4]
```diff
@@ -16,15 +17,38 @@ import (
 const REMOTE_CONFIG_CHECK_TIME_KEY = "REMOTE_CONFIG_CHECK_TIME"
 const REMOTE_CONFIG_CHECK_CYCLE_KEY = "REMOTE_CONFIG_CHECK_CYCLE"
 
+// store some metadata about the configuration settings
+//
```
[the remark this answers is not in the document]
Comment (bhou-crto) [reply internal/config/load.go]: Fixed

File: internal/config/load.go (hunk 3 above)
Comment (bhou-crto) [reply internal/config/load.go]: Fixed

**Review by lhutyra — APPROVED** [review]

**Review by gmonceyron — APPROVED** [review]

**Pull request merged**